File: 20221004_Tuesday/20221004.go
// Leetcode: 112. Path Sum.
// https://leetcode.com/problems/path-sum/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 0 ms, faster than 100.00% of Go online submissions for Path Sum.
// Memory Usage: 4.6 MB, less than 93.95% of Go online submissions for Path Sum.
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func hasPathSum(root *TreeNode, targetSum int) bool {
  if nil ==  root { return false}
  value := targetSum - root.Val
  if 0 == value && nil == root.Left && nil == root.Right {
    return true
  }
  if (nil != root.Left) && (nil !=root.Right){
    return hasPathSum(root.Left, value) || hasPathSum(root.Right, value)
  }else if (nil != root.Left){
    return hasPathSum(root.Left, value)
  }else if (nil != root.Right){
    return  hasPathSum(root.Right, value)
  } else {
    return false
  }
  return false
}

File: 20221004_Tuesday/20221004.rb
# Leetcode: 112. Path Sum.
# https://leetcode.com/problems/path-sum/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 188 ms, faster than 12.41% of Ruby online submissions for Path Sum.
# Memory Usage: 211.4 MB, less than 17.52% of Ruby online submissions for Path Sum.
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} target_sum
# @return {Boolean}
def has_path_sum(root, target_sum)
  return false if ! root
  value = target_sum - root.val
  if 0 == value && root.left.nil? && root.right.nil?
    return true
  end
  if root.left && root.right
    return has_path_sum(root.left,value) || has_path_sum(root.right, value)
  elsif root.left
    return has_path_sum(root.left,value)
  elsif root.right
    return has_path_sum(root.right,value)
  else
    return false # Leaf and value is not zero.
  end
  raise "Unreachable point."
end
